osu.Game/Overlays/Chat/ChatRecentTextBox.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics.UserInterface;
+using osuTK.Input;
+
+namespace osu.Game.Overlays.Chat
+{
+    public class ChatRecentTextBox : FocusedTextBox
+    {
+        private readonly List<string> messageHistory = new List<string>();
+
+        private int messageIndex = -1;
+
+        private string originalMessage = string.Empty;
+
+        protected override bool OnKeyDown(KeyDownEvent e)
+        {
+            /* Behavior:
+             * add when on last element -> last element stays
+             * subtract when on first element -> sets to original text
+             * reset indexing when Text is set to Empty
+             */
+
+            switch (e.Key)
+            {
+                case Key.Up:
+                    if (messageIndex == -1)
+                        originalMessage = Text;
+
+                    if (messageIndex == messageHistory.Count - 1)
+                        return true;
+
+                    Text = messageHistory[++messageIndex];
+
+                    return true;
+
+                case Key.Down:
+                    if (messageIndex == -1)
+                        return true;
+
+                    if (messageIndex == 0)
+                    {
+                        messageIndex = -1;
+                        Text = originalMessage;
+                        return true;
+                    }
+
+                    Text = messageHistory[--messageIndex];
+
+                    return true;
+            }
+
+            bool onKeyDown = base.OnKeyDown(e);
+
+            if (string.IsNullOrEmpty(Text))
+                messageIndex = -1;
```

Comment:
Will do that. I'm also thinking if it would be better to implement `IncrementIndex` and `DecrementIndex` methods for the inheritor to call when a specific key (`Key.Up` and `Key.Down` for now) is pressed.

Reply:
This is still unresolved, I think? I'm not actually sure I can tell due to the force-push.

Reply:
This has been already resolved by using `Current.ValueChanged`